File: Algo/src/main/java/Chapter4/Chapter43/Research/UnionFind.cpp
class UnionFind {
private:
    int *sets;
    int *size;
    int N;
public:
    UnionFind(int n) {
        // Set up a union-find data structure with n elements.
        N = n;
        sets = new int[N];
        size = new int[N];
        for (int i = 0 ; i < N; ++i) {
            sets[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;

        while (parent[root] != root) {
            parent[root] = parent[parent[root]]; // PATH COMPRESSION
            root = parent[root];
        }

        return root;
    }

    void merge(int x, int y) {
        int root_x = find(x);
        int root_y = find(y);

        if (size[root_y] > size[root_x]) {
            parent[root_x] = root_y;
            size[root_y] += size[root_x];
        } else {
            parent[root_y] = root_x;
            size[root_x] += size[root_y];
        }
    }
}

File: Algo/src/main/java/Chapter4/Chapter43/Research/UnionFind.java
package Chapter4.Chapter43.Research;

public class UnionFind {
    public int[] parent;
    public int[] size;

    public UnionFind(int N) {
        this.parent = new int[N];
        this.size = new int[N];

        for (int i = 0; i < N; ++i) {
            this.parent[i] = i;
            this.size[i] = 1;
        }
    }

    public int getRoot(int x) {
        while (this.parent[x] != x) {
            this.parent[x] = this.parent[this.parent[x]]; // path compression
            x = this.parent[x];
        }

        return x;
    }

    public boolean sameGroup(int a, int b) {
        // Returns true if a and b are in the same group, false otherwise.
        return this.getRoot(a) == this.getRoot(b);
    }

    public void union(int a, int b) {
        // Smaller tree goes into larger tree.
        int root_a = this.getRoot(a);
        int root_b = this.getRoot(b);

        if (size[root_a] > size[root_b]) {
            parent[root_b] = root_a;
            size[root_a] += size[root_b];
        } else {
            parent[root_a] = root_b;
            size[root_b] += size[root_a];
        }
    }

    private void print() {
        for (int i : this.parent) {
            System.out.print(i + " ");
        }
        System.out.println();
        for (int i : this.size) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        UnionFind uf = new UnionFind(5);

        uf.print();
        System.out.println();
        uf.union(0, 1);
        uf.union(2, 1);
        uf.union(3, 4);
        uf.print();

        System.out.println(uf.sameGroup(0, 1));
        System.out.println(uf.sameGroup(2, 1));
        System.out.println(uf.sameGroup(3, 4));
        System.out.println(uf.sameGroup(3, 0));
    }
}
